py-data-manipulation.Rmd

```diff
@@ -0,0 +1,457 @@
+```
+
+You can use logical operations to filter your data based on some criteria.
+
+```python
+# get rows where 'animal_gender' is female
+dog_licenses[dog_licenses['animal_gender'] == 'F'].head()
+
+# you can also get specific columns with .loc:
+dog_licenses.loc[dog_licenses['animal_gender'] == 'F', ['animal_name', 'animal_gender']].head()
+```
+
+A single expression can also be used to filter for several criteria, either
+matching all criteria (`&`) or any criteria (`|`). These special operators are
+used instead of `and` and `or` to make sure that the comparison occurs for each
+row in the data frame. Parentheses are added to indicate the priority of the
+comparisons.
+
+
+```python
+# the symbol '&' can be used to combine multiple logical comparisons
+# '==' means 'is equal to'
+dog_licenses.loc[(dog_licenses['animal_gender'] == 'F')
+         & (dog_licenses['breed_name'] == 'Boxer')
+           & (dog_licenses['animal_name'] == 'MILEY'), 'animal_name' : 'breed_name']
+
+# OR (|)
+# the order of comparisons can be structured with parentheses. Here we want the OR comparison
+# to happen before the AND comparison.
+
+dog_licenses.loc[(dog_licenses['animal_name'] == 'MILEY') & ((dog_licenses['breed_name'] == 'Boxer') |  
+          (dog_licenses['breed_name'] == 'Labrador Retriever')),  'animal_name' : 'breed_name']
+```
+
+## How can I calculate new values?
+
+A new column can be added to a data frame by calling the new column name and
+assigning values to it. The values should be either the same length as the
+original data frame, or all the same value.
+
+```python
+# create a new column and fill all rows with the value 'test'
+dog_licenses['new_column'] = 'test'
+dog_licenses.head()
+
+# drop the column we just made
+dog_licenses = dog_licenses.drop('new_column', axis = 1)
+```
+
+We can also create new columns by doing operations on existing columns.
+For example, let's create a column with the length of the dog's name.
+
+```python
+# the method .str allows string functions to be applied to the column contents (here we use 'len()')
+name_lengths = dog_licenses['animal_name'].str.len()
+
+# create a new column called 'name_length'
+dog_licenses['name_length'] = name_lengths
+
+# we can see that the new column has been added to the end of the data frame
+dog_licenses.columns
+```
+
+Let's create a column with the age of each animal at the time it was licensed.
+Note: this step takes a few seconds to run.
+
+```python
+# use to_datetime to create a datetime object for each date we're using
+animal_birth_date = pd.to_datetime(dog_licenses['animal_birth_month'])
+license_issued_datetime = pd.to_datetime(dog_licenses['license_issued_date'])
+
+# subtract the birth date from the license date
+age_when_licensed = license_issued_datetime - animal_birth_date
+
+# assign age_when_licensed to a new column
+dog_licenses['age_when_licensed'] = age_when_licensed
+```
+
+## How can I tell what's gone wronge in my programs?
```

Comment:
I feel this section should be in it's own chapter or somewhere else, as it doesn't really belong in "data wrangling".

Reply:
I think for now we leave as is, merge this asap, and figure out arranging sections at the editing phase.